lib/expand/join.go

```diff
@@ -860,3 +858,18 @@ func watchReconnects(ctx context.Context, cancel context.CancelFunc, watchCh <-c
 		}
 	}()
 }
+
+func validateConnection(ctx context.Context, agent *rpcserver.PeerServer) error {
+	connectTimeoutS := utils.GetenvWithDefault(defaults.EnvPeerConnectTimeout, defaults.PeerConnectTimeout.String())
+	connectTimeout, err := time.ParseDuration(connectTimeoutS)
+	if err != nil {
+		return trace.Wrap(err, "invalid agent connect timeout %q", connectTimeoutS)
+	}
+	ctx, cancel := context.WithTimeout(ctx, connectTimeout)
+	defer cancel()
+	err = agent.ValidateConnection(ctx)
+	if err != nil && !trace.IsLimitExceeded(err) {
```

Comment:
Not sure I understand this: so if the "validate" fails with timeout, we return "nil" as if it succeeded anyway?

Reply:
Yes, because it could still be connecting for all we know. There should be a separate (HTTP) timeout to set on all HTTP requests in general - out of scope here.

Reply:
Do we have any idea why the connection is timing out? Seems odd to me that validating connectivity to the server can really timeout, or as in the original report, that this is somehow related to the number of nodes joining the cluster.